### PR TITLE
Add Rails/PluralizationGrammar cop

### PR DESCRIPTION
**Pull request by maxjacobson**
This is just a pet peeve of mine. Not sure if it's worth including or not, but maybe? :smile:

### CODE REVIEW

File: spec/rubocop/cop/rails/pluralization_grammar_spec.rb [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        it 'does not register an offense' do
+          expect(cop.offenses).to be_empty
+        end
+      end
+
+      context "when using the singular form ##{method_name}" do
+        let(:source) { "some_method.#{method_name}" }
+
+        it 'registers an offense' do
+          expect(cop.offenses).to be_empty
```
Comment (jonas054) [inline spec/rubocop/cop/rails/pluralization_grammar_spec.rb]: The description says "registers an offense", and I do think it makes more sense to require plural forms here. `x.days.ago` etc.

File: spec/rubocop/cop/rails/pluralization_grammar_spec.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        end
+      end
+    end
+
+    [-1, -1.0, 1, 1.0].each do |singular_number|
```
Comment (rrosenblum) [inline spec/rubocop/cop/rails/pluralization_grammar_spec.rb]: Take a look at `TextUtil#pluralize`. Internally, RuboCop says that `-1` should be pluralized.

I am having trouble finding the comments in the PR talking about why we went that way, but I do remember doing a bit of research around what is grammatically correct. I think we found either is acceptable, but plural is preferred.

This seems to generate much debate based on context and logical grouping http://english.stackexchange.com/questions/9735/is-1-singular-or-plural.

From the sound of it, it comes down to how you interpret it `(-1.days)` vs `-(1.day)`

File: spec/rubocop/cop/rails/pluralization_grammar_spec.rb (hunk 1 above)
Comment (maxjacobson) [inline spec/rubocop/cop/rails/pluralization_grammar_spec.rb]: ah good catch. Indeed, at first, that's how I was implementing it, but it felt a bit complex, and then I started asking myself "well, what if the variable is called `one` and then RuboCop complains that the code says `one.day.ago` that wouldn't be great". I'll update the spec description for sure, but I'm inclined to leave the behavior to only lint for cases that are clearly wrong. How strongly do you feel about that suggestion?

File: spec/rubocop/cop/rails/pluralization_grammar_spec.rb (hunk 2 above)
Comment (maxjacobson) [inline spec/rubocop/cop/rails/pluralization_grammar_spec.rb]: that is *so* interesting. thanks for the explanation. I'll update it

File: spec/rubocop/cop/rails/pluralization_grammar_spec.rb (hunk 1 above)
Comment (jonas054) [inline spec/rubocop/cop/rails/pluralization_grammar_spec.rb]: That's a pretty weird special case. Who would create a variable or method called `one`? If you really want to handle it, make a special case in the code for receivers called `one`.

Comment (maxjacobson) [inline spec/rubocop/cop/rails/pluralization_grammar_spec.rb]: Yeah, I agree it's a contrived example :sweat_smile:. But I'm still not sure why RuboCop should have an opinion about the pluralization of unknown variables/methods. I suppose because when you have an unknown value, it's very likely to be pluralized?

File: lib/rubocop/cop/rails/pluralization_grammar.rb [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+        def plural_method?(method_name)
+          method_name.to_s.end_with?('s')
+        end
+
+        def singular_method?(method_name)
+          !plural_method?(method_name)
+        end
+
+        def singular_receiver?(receiver)
```
Comment (rrosenblum) [inline lib/rubocop/cop/rails/pluralization_grammar.rb]: I think that you could refactor this to take in the `number` instead of the `receiver`. The same would need to be done for `plural_receiver?`.

File: lib/rubocop/cop/rails/pluralization_grammar.rb [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        def literal_number?(node)
+          node.int_type? || node.float_type?
+        end
+
+        def pluralize(method_name)
+          method_name.to_s + 's'
+        end
+
+        def singularize(method_name)
+          method_name.to_s[0..-2]
```
Comment (rrosenblum) [inline lib/rubocop/cop/rails/pluralization_grammar.rb]: It looks like the call to `.to_s` is not needed here. Symbol implements `#[]` http://ruby-doc.org/core-1.9.3/Symbol.html#method-i-5B-5D.

File: lib/rubocop/cop/rails/pluralization_grammar.rb [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        def plural_receiver?(receiver)
+          !singular_receiver?(receiver)
+        end
+
+        def literal_number?(node)
+          node.int_type? || node.float_type?
+        end
+
+        def pluralize(method_name)
+          method_name.to_s + 's'
```
Comment (rrosenblum) [inline lib/rubocop/cop/rails/pluralization_grammar.rb]: Should this use string interpolation?
```ruby
"#{method_name}s"
```

File: lib/rubocop/cop/rails/pluralization_grammar.rb [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+      # core extensions to the numeric classes
+      #
+      # @example
+      #   # bad
+      #   3.day.ago
+      #
+      #   # good
+      #   45.minutes.from_now
+      class PluralizationGrammar < Cop
+        DURATION_METHODS = [:second, :seconds,
```
Comment (rrosenblum) [inline lib/rubocop/cop/rails/pluralization_grammar.rb]: I don't know how I feel about this being an array. It works well for the flat lookup of is this method one we care about, but having a mapping of the singular method to the plural method (and vice versa) feels like it would be better than adding or removing an 's' to singularize and pluralize.

File: spec/rubocop/cop/rails/pluralization_grammar_spec.rb [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+          let(:source) { "#{singular_number}.#{method_name}" }
+          it 'does not register an offense' do
+            expect(cop.offenses).to be_empty
+          end
+        end
+      end
+    end
+
+    context "when #{method_name} is called on any other literal number" do
+      [-rand(2..1000), -1, 0, rand(2..1000)].each do |plural_number|
```
Comment (rrosenblum) [inline spec/rubocop/cop/rails/pluralization_grammar_spec.rb]: I think some examples for decimal values would be nice to have.
```ruby
rand(-1.0...0)), rand(0...1.0)
```

File: lib/rubocop/cop/rails/pluralization_grammar.rb (hunk 3 above)
Comment (maxjacobson) [inline lib/rubocop/cop/rails/pluralization_grammar.rb]: :+1:

File: lib/rubocop/cop/rails/pluralization_grammar.rb (hunk 4 above)
Comment (maxjacobson) [inline lib/rubocop/cop/rails/pluralization_grammar.rb]: ah, did not realize. neat

File: lib/rubocop/cop/rails/pluralization_grammar.rb (hunk 5 above)
Comment (maxjacobson) [inline lib/rubocop/cop/rails/pluralization_grammar.rb]: I like it

**Comment by maxjacobson** [comment]
@rrosenblum thanks for the comments, I made some changes

File: CHANGELOG.md [hunk 8]
```diff
@@ -13,6 +13,7 @@
 * Allow `StyleGuide` parameters in local configuration for all cops, so users can add references to custom style guide documents. ([@cornelius][])
 * `UnusedMethodArgument` cop allows configuration to skip keyword arguments. ([@apiology][])
 * [#2318](https://github.com/bbatsov/rubocop/pull/2318): `Lint/Debugger` cop now checks for `Pry.rescue`. ([@rrosenblum][])
+* Add `Rails/PluralizationGrammar` cop, checks for incorrect grammar when using methods like 3.day.ago, when you should write 3.days.ago. ([@maxjacobson][])
```
Comment (rrosenblum) [inline CHANGELOG.md]: `3.day.ago` and `3.days.ago` should be in backticks.

File: config/enabled.yml [hunk 9]
```diff
@@ -1169,6 +1169,10 @@ Rails/Output:
   Description: 'Checks for calls to puts, print, etc.'
   Enabled: true
 
+Rails/PluralizationGrammar:
+  Description: 'Checks for incorrect grammar when using methods like 3.day.ago'
```
Comment (rrosenblum) [inline config/enabled.yml]: Same with here

File: lib/rubocop/cop/rails/pluralization_grammar.rb [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        DURATION_METHODS = { second: :seconds,
+                             minute: :minutes,
+                             hour: :hours,
+                             day: :days,
+                             week: :weeks,
+                             fortnight: :fortnights,
+                             month: :months,
+                             year: :years }
+
+        MSG = 'Prefer %s.%s for grammar reasons.'
```
Comment (rrosenblum) [inline lib/rubocop/cop/rails/pluralization_grammar.rb]: Same with around `%s.%s`.

File: lib/rubocop/cop/rails/pluralization_grammar.rb [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+          end
+        end
+
+        private
+
+        def autocorrect(node)
+          lambda do |corrector|
+            method_name = node.loc.selector.source
+            replacement = if plural_method?(method_name)
+                            singularize(method_name).to_s
```
Comment (rrosenblum) [inline lib/rubocop/cop/rails/pluralization_grammar.rb]: `singularize` and `pluralize` should handle the calls to `to_s` internally.

File: lib/rubocop/cop/rails/pluralization_grammar.rb [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        def literal_number?(node)
+          node.int_type? || node.float_type?
+        end
+
+        def pluralize(method_name)
+          DURATION_METHODS.fetch(method_name.to_sym)
+        end
+
+        def singularize(method_name)
+          DURATION_METHODS.invert.fetch(method_name.to_sym)
```
Comment (rrosenblum) [inline lib/rubocop/cop/rails/pluralization_grammar.rb]: I don't think I have ever used `invert` before. I am going to have to remember that one.

Performance wise, it seems to take significantly longer to have to `invert` the hash. It would be a good idea to extract `DURATION_METHODS.invert` to a constant.

```
INVERT = DURATION_METHODS.invert
Benchmark.ips do |x|
  x.report('regular') { DURATION_METHODS.fetch(:day) }
  x.report('invert') { DURATION_METHODS.invert.fetch(:days) }
  x.report('const') { INVERT.fetch(:days) }
  x.compare!
end  
Calculating -------------------------------------
             regular   140.330k i/100ms
              invert    34.673k i/100ms
               const   142.347k i/100ms
-------------------------------------------------
             regular      8.298M (± 4.3%) i/s -     41.397M
              invert    451.566k (±27.3%) i/s -      2.080M
               const      8.329M (± 4.3%) i/s -     41.565M

Comparison:
               const:  8328753.3 i/s
             regular:  8297891.3 i/s - 1.00x slower
              invert:   451566.4 i/s - 18.44x slower
```

**Comment by rrosenblum** [comment]
The commits are going to need to be squashed.

I like this cop. Being able to enforce grammar will help with readability.

File: lib/rubocop/cop/rails/pluralization_grammar.rb (hunk 12 above)
Comment (maxjacobson) [inline lib/rubocop/cop/rails/pluralization_grammar.rb]: gnarly

Yea, it's a handy method, but I suppose shouldn't be done over and over :sweat_smile:

**Comment by maxjacobson** [comment]
@rrosenblum pushed a few more commits (https://github.com/bbatsov/rubocop/commit/e12b81b24ae047bcb2e826f1cddfcda4fa12a3b7, https://github.com/maxjacobson/rubocop/commit/4e75f0af4b1a3b57a1f518a43bcc48c9647f6815, https://github.com/bbatsov/rubocop/commit/3d4604eeae0243ae6b618d17fd07b7c9f78a5e3a), and then squashed

File: lib/rubocop/cop/rails/pluralization_grammar.rb [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+
+module RuboCop
+  module Cop
+    module Rails
+      # This cop checks for correct grammar when using active support's
+      # core extensions to the numeric classes
+      #
+      # @example
+      #   # bad
+      #   3.day.ago
```
Comment (rrosenblum) [inline lib/rubocop/cop/rails/pluralization_grammar.rb]: Include an example for `1.days`.

File: lib/rubocop/cop/rails/pluralization_grammar.rb [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+                        :expression,
+                        format(MSG, number, pluralize(method_name)))
+          end
+        end
+
+        private
+
+        def autocorrect(node)
+          lambda do |corrector|
+            method_name = node.loc.selector.source
```
Comment (rrosenblum) [inline lib/rubocop/cop/rails/pluralization_grammar.rb]: I don't think that you need to retrieve the `source` here. Calling `node.loc.selector` will return the method as a symbol allowing `pluralize` and `singularize` to do their lookups without converting `method_name` back to a symbol. `plural_method?` will work as designed because of its call to `to_s`.

File: spec/rubocop/cop/rails/pluralization_grammar_spec.rb [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+        let(:source) { "some_method.#{method_name}" }
+
+        it 'does not register an offense' do
+          expect(cop.offenses).to be_empty
+        end
+      end
+    end
+
+    [1, 1.0].each do |singular_number|
+      context "when #{method_name} is called on a literal #{singular_number}" do
```
Comment (rrosenblum) [inline spec/rubocop/cop/rails/pluralization_grammar_spec.rb]: This outer context does not make sense to me. It is immediately followed by the inverse making it confusing what is being tested. Possibly shorten this to `'singular literals'` or expanding the inner `context` to `"when calling the plural method #{method_name}s on a the singular #{singular_number}"`.

**Comment by rrosenblum** [comment]
I found a couple of minor tweaks. Aside from that, it looks good to me. It will still need to be looked over by @jonas054 and @bbatsov

File: lib/rubocop/cop/rails/pluralization_grammar.rb (hunk 14 above)
Comment (maxjacobson) [inline lib/rubocop/cop/rails/pluralization_grammar.rb]: `node.loc.selector` returns an instance of `Parser::Source::Range`

Comment (rrosenblum) [inline lib/rubocop/cop/rails/pluralization_grammar.rb]: Whoops, this is fine then.

File: lib/rubocop/cop/rails/pluralization_grammar.rb [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+                                      hour: :hours,
+                                      day: :days,
+                                      week: :weeks,
+                                      fortnight: :fortnights,
+                                      month: :months,
+                                      year: :years }
+
+        PLURAL_DURATION_METHODS = SINGULAR_DURATION_METHODS.invert
+
+        MSG = 'Prefer `%s.%s` for grammar reasons.'
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/pluralization_grammar.rb]: I'd suggest `Prefer `%s.%s` over `%s.%s`.`. I believe further explanation here is redundant.

File: config/enabled.yml [hunk 17]
```diff
@@ -1169,6 +1169,10 @@ Rails/Output:
   Description: 'Checks for calls to puts, print, etc.'
   Enabled: true
 
+Rails/PluralizationGrammar:
+  Description: 'Checks for incorrect grammar when using methods like `3.day.ago`'
```
Comment (bbatsov) [inline config/enabled.yml]: This sentence should end with a `.`.

File: lib/rubocop/cop/rails/pluralization_grammar.rb [hunk 18]
```diff
@@ -0,0 +1,97 @@
+# encoding: utf-8
+
+module RuboCop
+  module Cop
+    module Rails
+      # This cop checks for correct grammar when using active support's
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/pluralization_grammar.rb]: ActiveSupport's

File: lib/rubocop/cop/rails/pluralization_grammar.rb [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+      # core extensions to the numeric classes
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/pluralization_grammar.rb]: Missing `.`

**Comment by bbatsov** [comment]
Apart from my small remarks - it looks OK. The commits have to be squashed and the commit message should adhere to our style.

**Comment by maxjacobson** [comment]
Thanks for the feedback, I've updated and squashed

**Comment by bbatsov** [comment]
:+1:
